fendermint/vm/topdown/src/sync/syncer.rs

```diff
@@ -0,0 +1,268 @@
+        if let Some((confirmed_height, payload)) = self.poll_next().await? {
+            atomically_or_err::<_, Error, _>(|| {
+                for h in (tail + 1)..confirmed_height {
+                    self.provider.new_parent_view(h, None)?;
+                    tracing::debug!(height = h, "null block pushed to cache");
+                }
+                self.provider
+                    .new_parent_view(confirmed_height, Some(payload.clone()))?;
+                tracing::debug!(height = confirmed_height, "non-null block pushed to cache");
+                Ok(())
+            })
+            .await?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<T, C> LotusParentSyncer<T, C>
+where
+    T: ParentFinalityStateQuery + Send + Sync + 'static,
+    C: tendermint_rpc::Client + Send + Sync + 'static,
+{
+    async fn is_syncing_peer(&self) -> anyhow::Result<bool> {
+        let status: tendermint_rpc::endpoint::status::Response = self
+            .tendermint_client
+            .status()
+            .await
+            .context("failed to get Tendermint status")?;
+        Ok(status.sync_info.catching_up)
+    }
+
+    /// Poll the next block height. Returns finalized and executed block data.
+    async fn poll_next(&mut self) -> Result<Option<(BlockHeight, ParentViewPayload)>, Error> {
+        let height = self.sync_pointers.head() + 1;
+        let parent_block_hash = self.non_null_parent_hash().await;
+
+        let block_hash_res = match self.parent_proxy.get_block_hash(height).await {
+            Ok(res) => res,
+            Err(e) => {
+                let err = e.to_string();
+                if is_null_round_str(&err) {
+                    tracing::warn!(height, "null round at height");
+
+                    self.sync_pointers.advance_head();
+
+                    return Ok(None);
+                }
+                return Err(Error::CannotQueryParent(err, height));
+            }
+        };
+
+        if block_hash_res.parent_block_hash != parent_block_hash {
+            tracing::warn!(
+                height,
+                parent_hash = hex::encode(&block_hash_res.parent_block_hash),
+                previous_hash = hex::encode(&parent_block_hash),
+                "parent block hash diff than previous hash",
+            );
+            return Err(Error::ParentChainReorgDetected);
+        }
+
+        if let Some(to_confirm) = self.sync_pointers.to_confirm() {
+            tracing::warn!(
+                height,
+                confirm = to_confirm,
+                "non-null round at height, confirmed previous height"
+            );
+            let data = self
+                .fetch_data(to_confirm, block_hash_res.block_hash)
+                .await?;
+            self.sync_pointers.advance_confirm(height);
+            return Ok(Some((to_confirm, data)));
+        }
+
+        tracing::warn!(height, "non-null round at height, waiting for confirmation");
+        self.sync_pointers.advance_head();
```

Comment:
I don't understand why the head doesn't move here :(

Reply:
It's an error, I fixed from integration testing.